### ISSUE TITLE
RemoteException causes IPython 8 to crash

### ISSUE THREAD

**Issue by jstucke** [ipython/ipython#13538]
Exceptions in subprocesses in `multiprocessing.pool.Pool` and `concurrent.futures.ProcessPoolExecutor` (and possibly more) cause IPython 8 to crash.

At first I thought this might be a duplicate of #12467 but while the symptoms are similar, the cause seems to be different

In `IPython.core.ultratb.VerboseTB.structured_traceback` exists a while loop which loops over the chain of exceptions. If it stems from a subprocess though, the type of the traceback might be `RemoteTraceback` (or `_RemoteTraceback` for `ProcessPoolExecutor`) instead of a regular traceback object which has neither a `__cause__` nor a `__context__`. This in turn leads to `etb` being `None` in `structured_traceback` and an exception being triggered by the newly introduced `assert etb is not None` in `format_exception_as_a_whole` which leads to the crash.

## Steps to reproduce:

```python
from multiprocessing.pool import Pool

def foo(x):
    raise Exception()

Pool(1).map(foo, [1])
```

Tested in latest Python 3.10 docker container with IPython 8.0.1:

<details>

```python
root@0d7056b87b09:/# ipython
Python 3.10.2 (main, Feb  8 2022, 04:44:29) [GCC 10.2.1 20210110]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.0.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from multiprocessing.pool import Pool

In [2]: def foo(x):
   ...:     raise Exception()

In [3]: Pool(1).map(foo, [1])
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "<ipython-input-2-f1f53f013b9a>", line 2, in foo
    raise Exception()
Exception
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3251, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-3-6128068d2903>", line 1, in <module>
    Pool(1).map(foo, [1])
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 364, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/local/lib/python3.10/multiprocessing/pool.py", line 771, in get
    raise self._value
Exception

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1934, in showtraceback
    stb = value._render_traceback_()
AttributeError: 'Exception' object has no attribute '_render_traceback_'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3191, in run_ast_nodes
    if await self.run_code(code, result, async_=asy):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3268, in run_code
    self.showtraceback(running_compiled_code=True)
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1936, in showtraceback
    stb = self.InteractiveTB.structured_traceback(etype,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 1105, in structured_traceback
    return FormattedTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 999, in structured_traceback
    return VerboseTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 871, in structured_traceback
    formatted_exceptions += self.format_exception_as_a_whole(etype, evalue, etb, lines_of_context,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 775, in format_exception_as_a_whole
    assert etb is not None
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1934, in showtraceback
    stb = value._render_traceback_()
AttributeError: 'AssertionError' object has no attribute '_render_traceback_'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 2814, in _run_cell
    return runner(coro)
  File "/usr/local/lib/python3.10/site-packages/IPython/core/async_helpers.py", line 129, in _pseudo_sync_runner
    coro.send(None)
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3012, in run_cell_async
    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3210, in run_ast_nodes
    self.showtraceback()
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1936, in showtraceback
    stb = self.InteractiveTB.structured_traceback(etype,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 1105, in structured_traceback
    return FormattedTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 999, in structured_traceback
    return VerboseTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 871, in structured_traceback
    formatted_exceptions += self.format_exception_as_a_whole(etype, evalue, etb, lines_of_context,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 775, in format_exception_as_a_whole
    assert etb is not None
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1934, in showtraceback
    stb = value._render_traceback_()
AttributeError: 'AssertionError' object has no attribute '_render_traceback_'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/ipython", line 8, in <module>
    sys.exit(start_ipython())
  File "/usr/local/lib/python3.10/site-packages/IPython/__init__.py", line 123, in start_ipython
    return launch_new_instance(argv=argv, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/traitlets/config/application.py", line 846, in launch_instance
    app.start()
  File "/usr/local/lib/python3.10/site-packages/IPython/terminal/ipapp.py", line 316, in start
    self.shell.mainloop()
  File "/usr/local/lib/python3.10/site-packages/IPython/terminal/interactiveshell.py", line 611, in mainloop
    self.interact()
  File "/usr/local/lib/python3.10/site-packages/IPython/terminal/interactiveshell.py", line 604, in interact
    self.run_cell(code, store_history=True)
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 2768, in run_cell
    result = self._run_cell(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 2819, in _run_cell
    self.showtraceback(running_compiled_code=True)
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1936, in showtraceback
    stb = self.InteractiveTB.structured_traceback(etype,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 1105, in structured_traceback
    return FormattedTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 999, in structured_traceback
    return VerboseTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 871, in structured_traceback
    formatted_exceptions += self.format_exception_as_a_whole(etype, evalue, etb, lines_of_context,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 775, in format_exception_as_a_whole
    assert etb is not None
AssertionError

If you suspect this is an IPython 8.0.1 bug, please report it at:
    https://github.com/ipython/ipython/issues
or send an email to the mailing list at ipython-dev@python.org

You can print a more detailed traceback right now with "%tb", or use "%debug"
to interactively debug it.

Extra-detailed tracebacks for bug-reporting purposes can be enabled via:
    %config Application.verbose_crash=True

root@0d7056b87b09:/# 
```

</details>

Also with `ProcessPoolExecutor`:

```python
from concurrent.futures import ProcessPoolExecutor

def foo():
  raise Exception("foo")

pool = ProcessPoolExecutor(max_workers=1)
future = pool.submit(foo)
future.result()
```

<details>

```python
root@8a840b5a619f:/# ipython
Python 3.10.2 (main, Feb  8 2022, 04:44:29) [GCC 10.2.1 20210110]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.0.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from concurrent.futures import ProcessPoolExecutor
   ...: 
   ...: 
   ...: def foo():
   ...:     raise Exception("foo")
   ...: 
   ...: 
   ...: pool = ProcessPoolExecutor(max_workers=1)
   ...: future = pool.submit(foo)
   ...: future.result()
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/concurrent/futures/process.py", line 243, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "<ipython-input-1-86d971dad3a2>", line 5, in foo
    raise Exception("foo")
Exception: foo
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3251, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-1-86d971dad3a2>", line 10, in <module>
    future.result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 446, in result
    return self.__get_result()
  File "/usr/local/lib/python3.10/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
Exception: foo

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1934, in showtraceback
    stb = value._render_traceback_()
AttributeError: 'Exception' object has no attribute '_render_traceback_'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3191, in run_ast_nodes
    if await self.run_code(code, result, async_=asy):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3268, in run_code
    self.showtraceback(running_compiled_code=True)
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1936, in showtraceback
    stb = self.InteractiveTB.structured_traceback(etype,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 1105, in structured_traceback
    return FormattedTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 999, in structured_traceback
    return VerboseTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 871, in structured_traceback
    formatted_exceptions += self.format_exception_as_a_whole(etype, evalue, etb, lines_of_context,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 775, in format_exception_as_a_whole
    assert etb is not None
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1934, in showtraceback
    stb = value._render_traceback_()
AttributeError: 'AssertionError' object has no attribute '_render_traceback_'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 2814, in _run_cell
    return runner(coro)
  File "/usr/local/lib/python3.10/site-packages/IPython/core/async_helpers.py", line 129, in _pseudo_sync_runner
    coro.send(None)
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3012, in run_cell_async
    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3210, in run_ast_nodes
    self.showtraceback()
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1936, in showtraceback
    stb = self.InteractiveTB.structured_traceback(etype,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 1105, in structured_traceback
    return FormattedTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 999, in structured_traceback
    return VerboseTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 871, in structured_traceback
    formatted_exceptions += self.format_exception_as_a_whole(etype, evalue, etb, lines_of_context,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 775, in format_exception_as_a_whole
    assert etb is not None
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1934, in showtraceback
    stb = value._render_traceback_()
AttributeError: 'AssertionError' object has no attribute '_render_traceback_'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/ipython", line 8, in <module>
    sys.exit(start_ipython())
  File "/usr/local/lib/python3.10/site-packages/IPython/__init__.py", line 123, in start_ipython
    return launch_new_instance(argv=argv, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/traitlets/config/application.py", line 846, in launch_instance
    app.start()
  File "/usr/local/lib/python3.10/site-packages/IPython/terminal/ipapp.py", line 316, in start
    self.shell.mainloop()
  File "/usr/local/lib/python3.10/site-packages/IPython/terminal/interactiveshell.py", line 611, in mainloop
    self.interact()
  File "/usr/local/lib/python3.10/site-packages/IPython/terminal/interactiveshell.py", line 604, in interact
    self.run_cell(code, store_history=True)
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 2768, in run_cell
    result = self._run_cell(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 2819, in _run_cell
    self.showtraceback(running_compiled_code=True)
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 1936, in showtraceback
    stb = self.InteractiveTB.structured_traceback(etype,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 1105, in structured_traceback
    return FormattedTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 999, in structured_traceback
    return VerboseTB.structured_traceback(
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 871, in structured_traceback
    formatted_exceptions += self.format_exception_as_a_whole(etype, evalue, etb, lines_of_context,
  File "/usr/local/lib/python3.10/site-packages/IPython/core/ultratb.py", line 775, in format_exception_as_a_whole
    assert etb is not None
AssertionError

If you suspect this is an IPython 8.0.1 bug, please report it at:
    https://github.com/ipython/ipython/issues
or send an email to the mailing list at ipython-dev@python.org

You can print a more detailed traceback right now with "%tb", or use "%debug"
to interactively debug it.

Extra-detailed tracebacks for bug-reporting purposes can be enabled via:
    %config Application.verbose_crash=True

root@8a840b5a619f:/# 
```

</details>

**Comment by Carreau** [comment]
I'll see if I can come up with a fix, but I can't reproduce. There are a lot of places that assume `exception.__traceback__` (etb) is not None.... so that might be tough to fix.

**Comment by jstucke** [comment]
> I'll see if I can come up with a fix, but I can't reproduce. 

That is weird. It should be easy to reproduce. A reliable way would be to use the python docker container:

```bash
docker pull python:3.10
docker run -it --rm --entrypoint=bash python:3.10
pip install ipython
ipython
In [1]: from multiprocessing.pool import Pool
   ...: def foo(x):
   ...:     raise Exception()
   ...: Pool(1).map(foo, [1])
```

I think we can agree that it isn't optimal that IPython crashes each time an exception inside a pool worker happens (this does not happen with IPython<8 and also works (kind of--at least it doesn't crash) when removing the new  `assert etb is not None`)

**Comment by jstucke** [comment]
I can confirm that #13588 fixes these crashes and also the ones #12467. Thank you!
